src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/TextMapPropagator.cs

```diff
@@ -0,0 +1,520 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text;
+using System.Net;
+using System.Collections.Generic;
+
+namespace System.Diagnostics
+{
+    public abstract class TextMapPropagator
+    {
+        public delegate bool PropagatorGetterCallback(object carrier, string fieldName, out string? value);
+
+        public abstract IReadOnlyCollection<string> Fields { get; }
+
+        // Inject
+
+        public abstract bool Inject(Activity activity, object carrier, Action<object /* carrier */, string /* field name */, string /* value to inject */> setter);
+        public abstract bool Inject(ActivityContext context, object carrier, Action<object /* carrier */, string /* field name */, string /* value to inject */> setter);
+        public abstract bool Inject(IEnumerable<KeyValuePair<string, string?>> baggage, object carrier, Action<object /* carrier */, string /* field name */, string /* value to inject */> setter);
+
+        // Extract
+
+        public abstract bool Extract(object carrier, PropagatorGetterCallback getter, out string? id, out string? state);
+        public abstract bool Extract(object carrier, PropagatorGetterCallback getter, out ActivityContext context);
+        public abstract bool Extract(object carrier, PropagatorGetterCallback getter, out IEnumerable<KeyValuePair<string, string?>>? baggage);
+
+        //
+        // Static APIs
+        //
+
+        public static TextMapPropagator DefaultPropagator { get; set; } = CreateLegacyPropagator();
+
+        // For Microsoft compatibility. e.g., it will propagate Baggage header name as "Correlation-Context" instead of "baggage".
+        public static TextMapPropagator CreateLegacyPropagator() => new LegacyTextMapPropagator();
+
+        // Suppress context propagation.
+        public static TextMapPropagator CreateOutputSuppressionPropagator() => new OutputSuppressionPropagator();
```

Comment:
I was chatting a bit with @shirhatti about locking down what kind of customization is in scope and what isn't. For example this propagator has legacy behavior for ingress and no egress. If someone wanted to only support TraceParent on ingress + no egress presumably they would need to write their own propagator. Likewise if someone wanted to egress TraceParent but not baggage they would also need their own propagator. I am fine with those customizations being out of scope, I just want to make sure everyone is in agreement and it aligns with what we've heard from customers.

Reply:
I agree there will be some scenarios users may need to write their own customized propagators. I would like to see first what these scenarios would be. If such scenarios truly little tweaks from what we are exposing, we still have the option to introduce more overloads to the static methods which can easily achieve such scenarios. something like `CreateOutputSuppressionPropagator(bool forceW3C)`.